java/runtime/src/main/java/io/ray/runtime/object/ObjectSerializer.java

```diff
@@ -86,8 +96,20 @@ public static NativeRayObject serialize(Object object) {
       return new NativeRayObject(serializedBytes, TASK_EXECUTION_EXCEPTION_META);
     } else {
       Pair<byte[], Boolean> serialized = Serializer.encode(object);
-      return new NativeRayObject(serialized.getLeft(), serialized.getRight() ?
-          OBJECT_METADATA_TYPE_CROSS_LANGUAGE : OBJECT_METADATA_TYPE_JAVA);
+      NativeRayObject nativeRayObject = new NativeRayObject(serialized.getLeft(),
+          serialized.getRight() ? OBJECT_METADATA_TYPE_CROSS_LANGUAGE : OBJECT_METADATA_TYPE_JAVA);
+      nativeRayObject.setContainedObjectIds(getAndClearContainedObjectIds());
```

Comment:
https://github.com/ray-project/ray/blob/5b192842b5f36ebac46a78ba4844ce019ac04343/python/ray/serialization.py#L194-L208

I'm not so sure. As the comment says, `self.is_in_band_serialization()` is false only in certain cases. In Java, users normally don't serialize object refs manually (in construct to invoking cloudpickle directly in Python). And I don't quite understand what "captured in a remote function/actor" means. 

One thing I missed is that when an exception occurs during serialization we should invoke `getAndClearContainedObjectIds` once in the catch block to clear the IDs.

Reply:
Ah okay, it's probably fine then.

Just for reference, "captured in a remote function/actor" means something like this:
```python
x = ray.put(...)
@ray.remote
def foo():
  ray.get(x)  # x is captured.
```
Not sure if this is even possible in java.

Reply:
I see. This is not a problem in Java because we don't serialize the function like Python do by pickling. We distribute user code via compiled jar files and invoke the methods by reflection. I haven't tested that if Ray Java API supports lambda expressions. But according to https://cr.openjdk.java.net/~briangoetz/lambda/lambda-translation.html, captured variables are additional parameters of the generated method (may vary based on different JDK implementations?). That makes it impossible to invoke remotely because we don't know the values of these additional parameters when trying to invoke it via reflection.